File: ticket_office/java/src/main/java/fr/arrolla/trainreservation/domain/Train.java
package fr.arrolla.trainreservation.domain;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Stream;

public class Train {
  private final HashMap<SeatID, Seat> seats = new HashMap<>();
  private final HashSet<CoachID> coaches = new HashSet<>();


  public Train(List<Seat> seats) {
    seats.forEach(seat -> {
      this.seats.put(seat.id(), seat);
      this.coaches.add(seat.coach());
    });
  }

  public Stream<Seat> seatsInCoach(CoachID coach) {
    return this.seats.values().stream().filter(seat -> seat.coach().equals(coach));
  }

  public Stream<Seat> seats() {
    return this.seats.values().stream();
  }

  public void reset() {
    seats.values().forEach(seat -> {
      seat.unBook();
    });
  }

  public void book(SeatID seatID, String bookingReference) {
    var seat = this.seats.get(seatID);
    if (seat == null) {
      String message = String.format("No seat with id '%s'", seatID.toString());
      throw new NoSuchSeatException(message);
    }
    seat.book(bookingReference);
  }

  public void applyReservation(Reservation reservation) {
    String bookingReference = reservation.bookingReference();
    reservation.seats().forEach(s -> book(s, bookingReference));
  }

  public double occupancyForCoach(CoachID coach) {
    var seats = seatsInCoach(coach).toList();
    var total = seats.size();
    var occupied = seats.stream().filter(Seat::isBooked).count();
    return occupied * 1.0f / total;
  }
}

File: python/ticket_office/ticket_office/domain/train.py
class Train:
    def __init__(self, seats):
        self._seats = {}
        self._coaches = set()
        for seat in seats:
            self._seats[seat.id] = seat
            self._coaches.add(seat.coach)

    @property
    def coaches(self):
        return sorted(self._coaches)

    def seats_in_coach(self, coach_id):
        return [s for s in self._seats.values() if s.coach == coach_id]

    def book(self, booking_reference, seat_ids):
        for seat_id in seat_ids:
            seat = self._seats.get(seat_id)
            if not seat:
                raise NoSuchSeat(seat_id)
            seat.book(booking_reference)

    def occupancy_after_booking(self, coach_id, seat_count):
        all_seats = self.seats_in_coach(coach_id)
        booked = [s for s in all_seats if s.is_booked()]
        return (len(booked) + seat_count) / len(all_seats) * 100


class NoSuchSeat(Exception):
    def __init__(self, seat_id):
        self.seat_id = seat_id
        super().__init__(seat_id)

    def __str__(self):
        return f"No such seat: {self.seat_id}"
